File: db4oj.tests/src/com/db4o/db4ounit/common/reflect/custom/CustomField.java
package com.db4o.db4ounit.common.reflect.custom;

import com.db4o.reflect.*;

/**
 * One important thing to remember when implementing ReflectField
 * is that getFieldType and getIndexType must always return ReflectClass
 * instances given by the parent reflector.
 */
public class CustomField implements ReflectField {

	// fields must be public so test works on less capable runtimes
	public CustomClassRepository _repository;	
	public String _name;
	public Class _type;
	public int _index;
	public boolean _indexed;

	public CustomField() {
	}

	public CustomField(CustomClassRepository repository, int index, String name, Class type) {
		_repository = repository;
		_index = index;
		_name = name;
		_type = type;
	}

	public Object get(Object onObject) {
		logMethodCall("get", onObject);
		return fieldValues(onObject)[_index];
	}

	private Object[] fieldValues(Object onObject) {
		return ((PersistentEntry)onObject).fieldValues;
	}

	public ReflectClass getFieldType() {
		logMethodCall("getFieldType");
		return _repository.forFieldType(_type);
	}

	public String getName() {
		return _name;
	}

	public Object indexEntry(Object orig) {
		logMethodCall("indexEntry", orig);
		return orig;
	}

	public ReflectClass indexType() {
		logMethodCall("indexType");
		return getFieldType();
	}

	public boolean isPublic() {
		return true;
	}

	public boolean isStatic() {
		return false;
	}

	public boolean isTransient() {
		return false;
	}

	public void set(Object onObject, Object value) {
		logMethodCall("set", onObject, value);
		fieldValues(onObject)[_index] = value;
	}

	public void indexed(boolean value) {
		_indexed = value;
	}
	
	public boolean indexed() {
		return _indexed;
	}

	public String toString() {
		return "CustomField(" + _index + ", " + _name + ", " + _type.getName() + ")";
	}

	private void logMethodCall(String methodName) {
		Logger.logMethodCall(this, methodName);
	}

	private void logMethodCall(String methodName, Object arg) {
		Logger.logMethodCall(this, methodName, arg);
	}

	private void logMethodCall(String methodName, Object arg1, Object arg2) {
		Logger.logMethodCall(this, methodName, arg1, arg2);
	}
}

File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Reflect/Custom/CustomField.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using Db4objects.Db4o.Reflect;
using Db4objects.Db4o.Tests.Common.Reflect.Custom;

namespace Db4objects.Db4o.Tests.Common.Reflect.Custom
{
	/// <summary>
	/// One important thing to remember when implementing ReflectField
	/// is that getFieldType and getIndexType must always return ReflectClass
	/// instances given by the parent reflector.
	/// </summary>
	/// <remarks>
	/// One important thing to remember when implementing ReflectField
	/// is that getFieldType and getIndexType must always return ReflectClass
	/// instances given by the parent reflector.
	/// </remarks>
	public class CustomField : IReflectField
	{
		public CustomClassRepository _repository;

		public string _name;

		public Type _type;

		public int _index;

		public bool _indexed;

		public CustomField()
		{
		}

		public CustomField(CustomClassRepository repository, int index, string name, Type
			 type)
		{
			// fields must be public so test works on less capable runtimes
			_repository = repository;
			_index = index;
			_name = name;
			_type = type;
		}

		public virtual object Get(object onObject)
		{
			LogMethodCall("get", onObject);
			return FieldValues(onObject)[_index];
		}

		private object[] FieldValues(object onObject)
		{
			return ((PersistentEntry)onObject).fieldValues;
		}

		public virtual IReflectClass GetFieldType()
		{
			LogMethodCall("getFieldType");
			return _repository.ForFieldType(_type);
		}

		public virtual string GetName()
		{
			return _name;
		}

		public virtual object IndexEntry(object orig)
		{
			LogMethodCall("indexEntry", orig);
			return orig;
		}

		public virtual IReflectClass IndexType()
		{
			LogMethodCall("indexType");
			return GetFieldType();
		}

		public virtual bool IsPublic()
		{
			return true;
		}

		public virtual bool IsStatic()
		{
			return false;
		}

		public virtual bool IsTransient()
		{
			return false;
		}

		public virtual void Set(object onObject, object value)
		{
			LogMethodCall("set", onObject, value);
			FieldValues(onObject)[_index] = value;
		}

		public virtual void Indexed(bool value)
		{
			_indexed = value;
		}

		public virtual bool Indexed()
		{
			return _indexed;
		}

		public override string ToString()
		{
			return "CustomField(" + _index + ", " + _name + ", " + _type.FullName + ")";
		}

		private void LogMethodCall(string methodName)
		{
			Logger.LogMethodCall(this, methodName);
		}

		private void LogMethodCall(string methodName, object arg)
		{
			Logger.LogMethodCall(this, methodName, arg);
		}

		private void LogMethodCall(string methodName, object arg1, object arg2)
		{
			Logger.LogMethodCall(this, methodName, arg1, arg2);
		}
	}
}
